File: 399.py
class Solution(object):
    def calcEquation(self, equations, values, queries):
        """
        :type equations: List[List[str]]
        :type values: List[float]
        :type queries: List[List[str]]
        :rtype: List[float]
        """
        graph = {}
        for (A,B), value in zip(equations,values):
            if A not in graph:
                graph[A] = {}
            if B not in graph:
                graph[B] = {}
            graph[A][B] = value
            graph[B][A] = 1./value

        def DFS(start,stop,visited):
            if start == stop:
                return 1.
            visited.add(start)
            for neighbor, value in graph[start].items():
                if neighbor not in visited:
                    result = DFS(neighbor,stop,visited)
                    if result != -1.:
                        return value * result
            return -1.

        return [DFS(C,D,set()) if C in graph and D in graph else -1. for C, D in queries]

"""
equations = [["a","b"],["b","c"]]
values = [2.0,3.0]
queries = [["a","c"],["b","a"],["a","e"],["a","a"],["x","x"]]
solution = Solution()
result = solution.calcEquation(equations, values, queries)
print(result)
"""

File: src/399/399.js
/**
 * @param {string[][]} equations
 * @param {number[]} values
 * @param {string[][]} queries
 * @return {number[]}
 */
var calcEquation = function(equations, values, queries) {
    // 1. Создаём граф как словарь смежности: ключ — вершина, значение — массив [сосед, вес]
    const graph = {};
    
    // 2. Заполняем граф на основе уравнений
    for (let i = 0; i < equations.length; i++) {
        const [a, b] = equations[i];
        const value = values[i];
        
        // Инициализируем вершины, если они ещё не существуют
        if (!graph[a]) graph[a] = [];
        if (!graph[b]) graph[b] = [];
        
        // Добавляем рёбра: a -> b с весом value, b -> a с весом 1/value
        graph[a].push([b, value]);
        graph[b].push([a, 1 / value]);
    }
    
    // 3. Функция DFS для поиска пути от start до end
    function dfs(start, end, visited) {
        // Если вершины нет в графе, возвращаем -1
        if (!graph[start] || !graph[end]) return -1.0;
        
        // Если достигли цели, возвращаем 1 (для случая start === end)
        if (start === end) return 1.0;
        
        visited.add(start);
        const neighbors = graph[start];
        
        // Обходим соседей
        for (const [next, weight] of neighbors) {
            if (!visited.has(next)) {
                const result = dfs(next, end, visited);
                // Если путь найден, возвращаем произведение весов
                if (result !== -1.0) {
                    return weight * result;
                }
            }
        }
        
        return -1.0;
    }
    
    // 4. Обрабатываем все запросы
    const results = [];
    for (const [c, d] of queries) {
        results.push(dfs(c, d, new Set()));
    }
    
    return results;
};

module.exports = calcEquation;